### PR TITLE
Improve circular data handling

### PR DESCRIPTION
**Pull request by gw000**
Right now trying to create circular data in Polar will cause a stack overflow due to unbounded recursion:

Screen Shot 2021-07-27 at 6 48 57 PM

This branch changes the dereferencing code to handle circular data a little more gracefully

Screen Shot 2021-07-27 at 6 55 33 PM

which still isn't super useful but probably better than crashing.

To manage this it does at least two slightly iffy things:

- It completely replaces `BindingManager::deref` with a slightly refactored `BindingManager::deep_deref`;
- To detect reference cycles it uses the same hash-set-of-hash-values-so-we-don't-have-to-clone strategy as the simplifier.

Local benchmarks don't seem to show much performance impact from deep-derefing everything, but someone else should test that also. A bigger concern would be if `deref` instead of `deep_deref` is actually necessary for some reason that's not covered by the tests.

### CODE REVIEW

File: polar-core/src/bindings.rs [hunk 1]
```diff
@@ -787,37 +780,6 @@ mod test {
         }
     }
 
-    #[test]
-    fn deref() {
```
Comment (dhatch) [inline polar-core/src/bindings.rs:791]: This test also covers some other logic in the binding manager, can we keep it but just replace `deref` with `deep_deref`?

File: polar-core/src/vm.rs [hunk 2]
```diff
@@ -719,7 +719,7 @@ impl PolarVirtualMachine {
     /// The exception is for lists, so that we can correctly handle rest variables.
     /// We also support cycle detection, in which case we return the original term.
     fn deref(&self, term: &Term) -> Term {
```
Comment (dhatch) [inline polar-core/src/vm.rs:721]: I don't think we need a `deref` function and a `deep_deref` function if they both do the same thing. What do you think?

**Review by dhatch — APPROVED** [review]
Small comments, lgtm. I don't need to review again after you address unless you have a question!

File: polar-core/src/bindings.rs (hunk 1 above)
Comment (gw000) [reply polar-core/src/bindings.rs:791]: 👍

File: polar-core/src/vm.rs (hunk 2 above)
Comment (gw000) [reply polar-core/src/vm.rs:721]: No, of course I agree. I'll do what you suggested and try to replace `deref` with `variable_state` where possible though.

Comment (gw000) [reply polar-core/src/vm.rs]: (there turned out to be only one such case)

**Pull request merged**